litt/disktable/segment/segment.go

```diff
@@ -0,0 +1,734 @@
+		metadata:          metadata,
+		keys:              keys,
+		shards:            shards,
+		keyFileSize:       keyFileSize,
+		deletionChannel:   make(chan struct{}, 1),
+	}
+
+	// Segments are returned with an initial reference count of 1, as the caller of the constructor is considered to
+	// have a reference to the segment.
+	segment.reservationCount.Store(1)
+
+	return segment, nil
+
+}
+
+// Size returns the size of the segment in bytes. Counts bytes that are on disk or that will eventually end up on disk.
+// This method is not thread safe, and should not be called concurrently with methods that modify the segment.
+func (s *Segment) Size() uint64 {
+	size := s.metadata.Size()
+
+	if s.IsSealed() {
+		// This segment is immutable, so it's thread safe to query the files directly.
+		size += s.keys.Size()
+		for _, shard := range s.shards {
+			size += shard.Size()
+		}
+	} else {
+		// This segment is mutable. We must use our local reckoning of the sizes of the files.
+		size += s.keyFileSize
+		for _, shardSize := range s.shardSizes {
+			size += shardSize
+		}
+	}
+
+	return size
+}
+
+// lookForFile looks for a file in a list of directories. It returns an error if the file appears
+// in more than one directory, and an empty string if the file is not found. If the file is found and
+// there are no errors, this method returns the path to the file.
+func lookForFile(directories []string, fileName string) (string, error) {
+	locations := make([]string, 0, 1)
+	for _, directory := range directories {
+		potentialLocation := path.Join(directory, fileName)
+		exists, err := util.Exists(potentialLocation)
+		if err != nil {
+			return "", fmt.Errorf("failed to check if file %s exists: %v", potentialLocation, err)
+		}
+		if exists {
+			locations = append(locations, potentialLocation)
+		}
+	}
+
+	if len(locations) > 1 {
+		return "", fmt.Errorf("file %s found in multiple directories: %v", fileName, locations)
+	}
+
+	if len(locations) == 0 {
+		return "", nil
+	}
+	return locations[0], nil
+}
+
+// SetNextSegment sets the next segment in the chain.
+func (s *Segment) SetNextSegment(nextSegment *Segment) {
+	nextSegment.Reserve()
+	s.nextSegment = nextSegment
+}
+
+// GetShard returns the shard number for a key.
+func (s *Segment) GetShard(key []byte) uint32 {
+	if s.metadata.shardingFactor == 1 {
+		// Shortcut: if we have one shard, we don't need to hash the key to figure out the mapping.
+		return 0
+	}
+
+	return util.HashKey(key, s.metadata.salt) % s.metadata.shardingFactor
+}
+
+// Write records a key-value pair in the data segment, returning the maximum size of all shards within this segment.
+//
+// This method does not ensure that the key-value pair is actually written to disk, only that it will eventually be
+// written to disk. Flush must be called to ensure that all data previously passed to Write is written to disk.
+func (s *Segment) Write(data *types.KVPair) (keyCount uint64, keyFileSize uint64, maxShardSize uint64, err error) {
```

Comment:
Returning the max shard size etc from a Write looks a bit odd, can they be a separate method?

Reply:
I've moved this data to a getter method `GetMaxShardSize()`.